src/libguac/parser.c

```diff
@@ -222,7 +222,11 @@ int guac_parser_read(guac_parser* parser, guac_socket* socket, int usec_timeout)
             retval = guac_socket_select(socket, usec_timeout);
             if (retval <= 0)
                 return -1;
-           
+            
+            /* Reset pointers if instruction buf len is less than max instruction len */
+            if (buffer_end - unparsed_end < GUAC_INSTRUCTION_MAX_LENGTH)
+                unparsed_end = unparsed_start = parser->__instructionbuf;
```

Comment:
This can result in the corruption of the in-progress instruction. There is no guarantee that the sequence of reads which trigger the `if` condition will not (for example) split part of an instruction element. While the normal pattern of updates to `unparsed_start` and `unparsed_end` would guarantee that the sequence of reads ultimately builds a contiguous instruction, that guarantee is lost if those pointers are updated in this way - reads are suddenly no longer contiguous.

Reply:
`GUAC_STATUS_NO_MEMORY` happens only if it is impossible to free up space within the buffer using `memmove()`, a fatal condition. The check `unparsed_end == buffer_end` is not about any individual `read()` call; it's about the union of all `read()` calls since the buffer was last reset.

Reply:
@changkun, if you are confident that these changes are correct, can you perhaps provide a sequence of `read()`s that demonstrate that the parser (without these changes) incorrectly generates a parse error?

Reply:
@mike-jumper All description lies in the jira ticket: https://issues.apache.org/jira/projects/GUACAMOLE/issues/GUACAMOLE-620

Frist, all TCP traffic contains full instruction.

Second, the following outputs:

```
__write_all: 4.sync,8.76258350;
__write_all: 5.mouse,3.445,3.261,1.1;
guac_socket_fd_read_handler: 4.sync,8.76258�, length: 14
guac_socket_fd_read_handler: 5.mouse,3.445,3.261,1.1;eo, length: 24
```

`__write_all` in `guacd` successfully writes full instruction, somehow (no idea with system calls behavior),  `guac_socket_fd_read_handler` lost part of the instruction, and then reads next instruction.

Reply:
> First, all TCP traffic contains full instruction.

What matters when considering the validity of these changes and the underlying cause of what you're seeing are the specific results of each call to `read()`. Each call to `read()` will populate the buffer with an unknown portion of that stream. The size of each chunk in the sequence matters.

If it's been established that the TCP stream is correct in itself, that's good, but that's not the whole picture.

> ... somehow (no idea with system calls behavior), `guac_socket_fd_read_handler` lost part of the instruction, and then reads next instruction.

That isn't possible. Each call to `read()` is not guaranteed to return all data written to the socket - it may only return part of that data, but the next call to `read()` should return yet more of the data, and so on. It's not possible for data to appear to just vanish between reads unless:

1. Multiple threads are both consuming the data from the same file descriptor.
2. `__write_all()` is not actually writing everything.

You are testing against unmodified guacamole-server master, correct? You're not including your changes from #179?

Reply:
> That isn't possible. Each call to read() is not guaranteed to return all data written to the socket - it may only return part of that data, but the next call to read() should return yet more of the data, and so on.

I know it is not possible, socketpair has its own buffer on both side of sockets, they cache unread data. Magically, I somehow meet this problem.

> __write_all() is not actually writing everything.

So far, I observed from the output log, `__write_all` write entire buffer array successfully, in the end, before parse error.

> You are testing against unmodified guacamole-server master, correct? You're not including your changes from #179?

No, I do not including #179 . All tests and tracking are based on vanilla tag 0.9.14.

Reply:
Here, I've added debug logging for each block written by `write()` within `__write_all()` as well as each block read via `guac_socket_read()` within the parser:

https://github.com/mike-jumper/guacamole-server/tree/debug-io

The logging added above should be stable in the sense that:

1. Reads and writes are written to separate files and thus the log shouldn't become corrupt if they happen to overlap.
2. The log routine does not expect a null terminator, relying instead on a supplied length, so there shouldn't be garbage / stale bytes in the logs that weren't actually read/written at the time logged.

Reads are logged to `read-log.txt` within the current directory of guacd and writes are logged to `write-log.txt` within the current directory of guacd. The logged data is in hex dump format and will generally look like this:

```
--- 25 bytes:
00000000:  34 2E 73 69 7A 65 2C 34 2E 31 31 35 34 2C 33 2E  4.size,4.1154,3.
00000010:  36 35 34 2C 32 2E 39 36 3B                       654,2.96;
--- 209 bytes:
00000000:  35 2E 61 75 64 69 6F 2C 38 2E 61 75 64 69 6F 2F  5.audio,8.audio/
00000010:  4C 38 2C 39 2E 61 75 64 69 6F 2F 4C 31 36 3B 35  L8,9.audio/L16;5
00000020:  2E 76 69 64 65 6F 3B 35 2E 69 6D 61 67 65 2C 31  .video;5.image,1
00000030:  30 2E 69 6D 61 67 65 2F 6A 70 65 67 2C 39 2E 69  0.image/jpeg,9.i
00000040:  6D 61 67 65 2F 70 6E 67 2C 31 30 2E 69 6D 61 67  mage/png,10.imag
00000050:  65 2F 77 65 62 70 3B 37 2E 63 6F 6E 6E 65 63 74  e/webp;7.connect
00000060:  2C 39 2E 6C 6F 63 61 6C 68 6F 73 74 2C 34 2E 35  ,9.localhost,4.5
00000070:  39 30 31 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C  901,0.,0.,0.,0.,
00000080:  30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30  0.,0.,0.,0.,0.,0
00000090:  2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E  .,0.,0.,0.,0.,0.
000000A0:  2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C  ,0.,0.,0.,0.,0.,
000000B0:  30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30  0.,0.,0.,0.,0.,0
000000C0:  2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E 2C 30 2E  .,0.,0.,0.,0.,0.
000000D0:  3B                                               ;
--- 30 bytes:
00000000:  39 2E 63 6C 69 70 62 6F 61 72 64 2C 31 2E 30 2C  9.clipboard,1.0,
00000010:  31 30 2E 74 65 78 74 2F 70 6C 61 69 6E 3B        10.text/plain;
--- 76 bytes:
00000000:  33 2E 65 6E 64 2C 31 2E 30 3B 35 2E 61 75 64 69  3.end,1.0;5.audi
00000010:  6F 2C 31 2E 30 2C 33 31 2E 61 75 64 69 6F 2F 4C  o,1.0,31.audio/L
00000020:  31 36 3B 72 61 74 65 3D 34 34 31 30 30 2C 63 68  16;rate=44100,ch
00000030:  61 6E 6E 65 6C 73 3D 32 3B 34 2E 73 79 6E 63 2C  annels=2;4.sync,
00000040:  39 2E 38 33 30 34 34 37 36 30 32 3B              9.830447602;
...
```

This should hopefully give us cleaner, more thorough logs.

Please try to reproduce the parse error with the above branch and provide the resulting `write-log.txt` and `read-log.txt` for the connection that failed. Be sure to remove the `write-log.txt` and `read-log.txt` files between connection attempts, as the debug code just repeatedly appends to those files. If the files exist at the beginning of a connection, the old log data will still be there.

Reply:
Log files are uploaded.

Reply:
Well, the good news is that your servlet is absolved. There can be no doubt that the protocol data is indeed reaching guacd correctly and is not being clobbered by shared use of a common socket as I originally suspected. [Your provided `write_log.txt`](https://issues.apache.org/jira/secure/attachment/12939404/write-log.txt) clearly demonstrates this:

```
...
--- 23 bytes:
00000000:  35 2E 6D 6F 75 73 65 2C 33 2E 35 34 31 2C 32 2E  5.mouse,3.541,2.
00000010:  32 36 2C 31 2E 31 3B                             26,1.1;
--- 23 bytes:
00000000:  35 2E 6D 6F 75 73 65 2C 33 2E 35 34 32 2C 32 2E  5.mouse,3.542,2.
00000010:  32 36 2C 31 2E 31 3B                             26,1.1;
```

The bad news is that these changes are definitely incorrect, and they do indeed break the parser. The following are the blocks ultimately read by simple, sequential calls to `guac_socket_read()` within the parser according to [your provided `read.txt`](https://issues.apache.org/jira/secure/attachment/12939405/read-log.txt):

```
...
--- 20 bytes:
00000000:  35 2E 6D 6F 75 73 65 2C 33 2E 35 34 31 2C 32 2E  5.mouse,3.541,2.
00000010:  32 36 2C 31                                      26,1
--- 23 bytes:
00000000:  35 2E 6D 6F 75 73 65 2C 33 2E 35 34 32 2C 32 2E  5.mouse,3.542,2.
00000010:  32 36 2C 31 2E 31 3B                             26,1.1;
```

Any changes which allow the parser to accept "5.mouse,3.541,2.26,15.mouse,3.542,2.26,1.1;" are incorrect, as that should result in a parse error. It is invalid protocol data.

The issue is not that the parser is behaving incorrectly - it is correctly bailing out with a parse error when given invalid data. The issue is that bytes are somehow vanishing from the stream between otherwise normal calls to `read()`.

Reply:
As expected and as same as what I wrote before.

Two suggestions if you do what make changes from this PR:

1. Carefully handling conditions when instruction incomplete, for example, reconstruction mechanism in a transmission context.
2. Redesigning protocol, add probably one instruction which allows `client/server` retransmission, as far as I see, there is no error handling for transmission. Also, `LENGTH.VALUE` is a redundant design and it can be simplified with `VALUE` only.